pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go

```diff
@@ -845,52 +845,64 @@ func (lbaas *LbaasV2) EnsureLoadBalancer(ctx context.Context, clusterName string
+		klog.V(4).Infof("Ensure an external loadbalancer service")
+
+		class := getStringFromServiceAnnotation(apiService, ServiceAnnotationLoadBalancerClass, "")
+		if class != "" {
+			lbClass = lbaas.opts.LBClasses[class]
+			if lbClass == nil {
+				return nil, fmt.Errorf("invalid loadbalancer class %q", class)
+			}
+
+			klog.V(4).Infof("found loadbalancer class %q with %+v", class, lbClass)
+
+			// read floating network id and floating subnet id from loadbalancer class
+			if lbClass.FloatingNetworkID != "" {
+				floatingNetworkID = lbClass.FloatingNetworkID
+			}
+
+			if lbClass.FloatingSubnetID != "" {
+				floatingSubnetID = lbClass.FloatingSubnetID
 			}
 		}
-		klog.V(4).Infof("found loadbalancer class %q with %+v", class, lbClass)
-	} else {
-		internalAnnotation, err = getBoolFromServiceAnnotation(apiService, ServiceAnnotationLoadBalancerInternal, lbaas.opts.InternalLB)
-		if err != nil {
-			return nil, err
-		}
-	}
 
-	floatingPool := getStringFromServiceAnnotation(apiService, ServiceAnnotationLoadBalancerFloatingNetworkID, lbaas.opts.FloatingNetworkID)
-	if lbClass != nil && lbClass.FloatingNetworkID != "" {
-		floatingPool = lbClass.FloatingNetworkID
-		klog.V(4).Infof("found floating network id %q from class %q", floatingPool, class)
-	}
-	if len(floatingPool) == 0 {
-		var err error
-		floatingPool, err = getFloatingNetworkIDForLB(lbaas.network)
-		if err != nil {
-			klog.Warningf("Failed to find floating-network-id for Service %s: %v", serviceName, err)
+		if floatingNetworkID == "" {
+			floatingNetworkID = getStringFromServiceAnnotation(apiService, ServiceAnnotationLoadBalancerFloatingNetworkID, lbaas.opts.FloatingNetworkID)
+			if floatingNetworkID == "" {
+				var err error
+				floatingNetworkID, err = getFloatingNetworkIDForLB(lbaas.network)
+				if err != nil {
+					klog.Warningf("Failed to find floating-network-id for Service %s: %v", serviceName, err)
+				}
+			}
 		}
-	}
 
-	switch internalAnnotation {
-	case true:
-		klog.V(4).Infof("Ensure an internal loadbalancer service.")
-	case false:
-		if len(floatingPool) != 0 {
-			klog.V(4).Infof("Ensure an external loadbalancer service, using floatingPool: %v", floatingPool)
-		} else {
-			return nil, fmt.Errorf("floating-network-id or loadbalancer.openstack.org/floating-network-id should be specified when ensuring an external loadbalancer service")
+		if floatingSubnetID == "" {
+			floatingSubnetName := getStringFromServiceAnnotation(apiService, ServiceAnnotationLoadBalancerFloatingSubnet, "")
+			if floatingSubnetName != "" {
+				lbSubnet, err := lbaas.getSubnet(floatingSubnetName)
+				if err != nil || lbSubnet == nil {
+					klog.Warningf("Failed to find floating-subnet-id for Service %s: %v", serviceName, err)
+				} else {
+					floatingSubnetID = lbSubnet.ID
+				}
+			}
+
+			if floatingSubnetID == "" {
+				floatingSubnetID = getStringFromServiceAnnotation(apiService, ServiceAnnotationLoadBalancerFloatingSubnetID, lbaas.opts.FloatingSubnetID)
```

Comment:
Can we check if the subnet ID is in the floating network?

Reply:
sure, that's a good check to make. I'll update the PR

Reply:
I have pushed a new [commit](https://github.com/kubernetes/cloud-provider-openstack/pull/1102/commits/039f611cc1c6d37a7e72b76408991e617d5e82f9) to check if subnet belongs to the network. Can you review again? Thanks